File: Server_Root/Server/RoomServer/Messages/GameEventTeleportToR.cpp
#include "stdafx.h"
#include "GameEventTeleportToR.h"
#include "GameEventTeleportToR_Data_generated.h"

GameEventTeleportToR::GameEventTeleportToR() : m_Builder(1024)
{
}

GameEventTeleportToR::~GameEventTeleportToR()
{
}

unsigned short GameEventTeleportToR::GetID()
{
	return MESSAGE_ID;
}

IMessage* GameEventTeleportToR::Clone()
{
	return NULL;
}

const char* GameEventTeleportToR::Serialize(int* pLength)
{
	FBSData::Vector3 start(m_vec3Start.x, m_vec3Start.y, m_vec3Start.z);
	FBSData::Vector3 dest(m_vec3Dest.x, m_vec3Dest.y, m_vec3Dest.z);

	GameEventTeleportToR_DataBuilder data_builder(m_Builder);
	data_builder.add_PlayerIndex(m_nPlayerIndex);
	data_builder.add_Start(&start);
	data_builder.add_Dest(&dest);
	data_builder.add_State(m_nState);
	auto data = data_builder.Finish();

	m_Builder.Finish(data);

	*pLength = m_Builder.GetSize();

	return (char*)m_Builder.GetBufferPointer();
}

bool GameEventTeleportToR::Deserialize(const char* pChar)
{
	auto data = flatbuffers::GetRoot<GameEventTeleportToR_Data>((const void*)pChar);

	m_nPlayerIndex = data->PlayerIndex();
	m_vec3Start.x = data->Start()->x();
	m_vec3Start.y = data->Start()->y();
	m_vec3Start.z = data->Start()->z();
	m_vec3Dest.x = data->Dest()->x();
	m_vec3Dest.y = data->Dest()->y();
	m_vec3Dest.z = data->Dest()->z();
	m_nState = data->State();

	return true;
}

File: Client_Root/Client/Assets/Scripts/Network/Messages/ToRoom/GameEventTeleportToR.cs
using UnityEngine;
using FlatBuffers;

public class GameEventTeleportToR : IMessage
{
    public const ushort MESSAGE_ID = MessageID.GameEventTeleportToR_ID;

    public int m_nPlayerIndex;
    public Vector3 m_vec3Start;
    public Vector3 m_vec3Dest;
    public int m_nState;

    public ushort GetID()
    {
        return MESSAGE_ID;
    }

    public IMessage Clone()
    {
        return null; 
    }

    public byte[] Serialize()
    {
        FlatBufferBuilder builder = new FlatBufferBuilder(1024);

        var start = FBSData.Vector3.CreateVector3(builder, m_vec3Start.x, m_vec3Start.y, m_vec3Start.z);
        var dest = FBSData.Vector3.CreateVector3(builder, m_vec3Dest.x, m_vec3Dest.y, m_vec3Dest.z);

        GameEventTeleportToR_Data.StartGameEventTeleportToR_Data(builder);
        GameEventTeleportToR_Data.AddPlayerIndex(builder, m_nPlayerIndex);
        GameEventTeleportToR_Data.AddStart(builder, start);
        GameEventTeleportToR_Data.AddDest(builder, dest);
        GameEventTeleportToR_Data.AddState(builder, m_nState);
        var data = GameEventTeleportToR_Data.EndGameEventTeleportToR_Data(builder);

        builder.Finish(data.Value);

        return builder.SizedByteArray();
    }

    public bool Deserialize(byte[] bytes)
    {
        var buf = new ByteBuffer(bytes);

        var data = GameEventTeleportToR_Data.GetRootAsGameEventTeleportToR_Data(buf);

        m_nPlayerIndex = data.PlayerIndex;
        m_vec3Start = new Vector3(data.Start.X, data.Start.Y, data.Start.Z);
        m_vec3Dest = new Vector3(data.Dest.X, data.Dest.Y, data.Dest.Z);
        m_nState = data.State;

        return true;
    }
}
